mmdet/core/hook/wandblogger_hook.py

```diff
@@ -0,0 +1,482 @@
+
+        Args:
+            path_to_model (str): Path where model checkpoints are saved.
+            epoch (int): The current epoch.
+            aliases (list): List of the aliases associated with this artifact.
+            metadata (dict, optional): Metadata associated with this artifact.
+        """
+        model_artifact = self.wandb.Artifact(
+            f'run_{self.wandb.run.id}_model', type='model', metadata=metadata)
+        model_artifact.add_file(f'{path_to_model}/epoch_{epoch+1}.pth')
+        self.wandb.log_artifact(model_artifact, aliases=aliases)
+
+    def _get_ckpt_metadata(self, runner):
+        """Get model checkpoint metadata."""
+        current_epoch = runner.epoch + 1
+
+        if (current_epoch % self.ckpt_hook.interval
+                == 0) and (current_epoch % self.eval_hook.interval == 0):
+            results = self.eval_hook.results
+
+            eval_results = self.val_dataset.evaluate(results, logger='silent')
+            metadata = dict(epoch=runner.epoch + 1, **eval_results)
+            return metadata
+        else:
+            return {}
+
+    def _init_data_table(self):
+        """Initialize the W&B Tables for validation data."""
+        columns = ['image_name', 'image']
+        self.data_table = self.wandb.Table(columns=columns)
+
+    def _init_pred_table(self):
+        """Initialize the W&B Tables for model evaluation."""
+        columns = ['epoch', 'image_name', 'ground_truth', 'prediction']
+        self.eval_table = self.wandb.Table(columns=columns)
+
+    def _add_ground_truth(self):
+        # Get image loading pipeline
+        from mmdet.datasets.pipelines import LoadImageFromFile
+        transforms = self.val_dataset.pipeline.transforms
+        for transform in transforms:
+            if isinstance(transform, LoadImageFromFile):
+                img_loader = transform
+        if 'img_loader' not in locals():
+            warnings.warn(
+                'LoadImageFromFile is required to add images '
+                'to W&B Tables.', UserWarning)
+            self.log_evaluation = False
+
+        # Determine the number of samples to be logged.
+        num_total_images = len(self.val_dataset)
+        if self.num_eval_images > num_total_images:
+            warnings.warn(
+                'The num_eval_images is greater than the total number '
+                'of validation samples. Complete validation set '
+                'will be logged.', UserWarning)
+        self.num_eval_images = min(self.num_eval_images, num_total_images)
+
+        classes = self.val_dataset.CLASSES
+        self.class_id_to_label = {
+            id + 1: name
+            for id, name in enumerate(classes)
+        }
+        self.class_set = self.wandb.Classes([{
+            'id': id,
+            'name': name
+        } for id, name in self.class_id_to_label.items()])
+
+        img_prefix = self.val_dataset.img_prefix
+
+        for idx in range(self.num_eval_images):
```

Comment:
replace:
```python

eval_image_indexs = np.arange(self.num_eval_images)
np.random.shuffle(eval_image_indexs)
for idx in eval_image_indexs:
      ....

Reply:
This is a better implementation. Thanks. :)